lib/goal.ex

```diff
@@ -793,6 +824,65 @@ defmodule Goal do
     end)
   end
 
+  defp validate_array_basic_field(changes, field, schema, type, changeset) do
+    params = Map.get(changes, field)
+    rules = Map.get(schema, field)
+    item_rules = Keyword.get(rules, :rules)
+
+    if item_rules && is_list(params) do
+      case reduce_and_validate_array_basic_fields(item_rules, type, field, params) do
+        {:valid, changes} ->
+          changeset
+          |> put_in([Access.key(:changes), Access.key(field)], Enum.reverse(changes))
+          |> Map.update!(:valid?, &Kernel.&&(&1, true))
+
+        {:invalid, errors} ->
+          changeset
+          |> Map.update!(:errors, &(errors ++ &1))
+          |> Map.put(:valid?, false)
+      end
+    else
+      changeset
+    end
+  end
+
+  @placeholder :inner
+  defp reduce_and_validate_array_basic_fields(schema, type, field, params) do
+    schema = %{@placeholder => [{:type, type} | schema]}
+
+    Enum.reduce(params, {:valid, []}, fn params, {status, acc} ->
+      changeset = build_changeset(schema, %{@placeholder => params})
+
+      case {status, changeset} do
+        {:valid, %Changeset{valid?: true, changes: %{@placeholder => inner_changes}}} ->
+          {:valid, [inner_changes | acc]}
+
+        {:valid, %Changeset{valid?: false, errors: errors}} ->
+          # The implementation is "all or nothing", so even if there where successful
+          # changes before, we reset them now since the whole changeset isn't valid.
+          {:invalid, errors}
+
+        {:invalid, %Changeset{valid?: false, errors: errors}} ->
+          {:invalid, errors ++ acc}
+
+        {:invalid, %Changeset{valid?: true}} ->
+          {:invalid, acc}
+      end
+    end)
+    |> case do
+      {:invalid, errors} ->
+        errors =
+          errors
+          |> Enum.uniq()
```

Comment:
`Enum.uniq/1` is there to remove any duplicates. These will naturally happen when for example `min: 3` is given as the rule and multiple items fail to meet it.

Instead of doing this, I have re-implemented it using `MapSet` instead.